### PR TITLE
Core update for solvers.

### PR DESCRIPTION
**Pull request by Ericgig**
**Description**
As proposed by @hodgestar, I am splitting #1409 into smaller PRs. This is part 1/3, but contain some change in core that would come later in the Solver PR:

- Add manual cython dispatched function. Used by vern ODE method.

- Add `to` to `QobjEvoFunc` and some check. Used by brmesolve rework.

- Add a version to string Coefficient, a 'try_parsing' options, partial arguments update and basic cleaning function.
This comes mostly from talk with @jakelishman.
When updating coefficient with new argument, it no longer require all coefficient to be given, but will simply update those available. 
Since those change must overwrite old version of the coefficient, a version tag is added to the coefficient folder and a function to erase the old version. It use it's own `COEFF_VERSION`, not qutip version since I don't expect most qutip update will not touch `coefficient.pyx` so we need not to force user to recompile. But it raise the risk of forgetting to change it when working on them. 
Lastly, I added an option to disable the string coefficient parsing since there seems to have some reservation in the last meeting.
But there is a weakness when using that form if unused args are passed: the `key = args['key']` code could be generated for those extra `args` but will not affect the hash name. 

**Related issues or PRs**
Will replace part of #1409.

### CODE REVIEW

File: qutip/tests/core/test_coefficient.py
Comment (hodgestar) [inline qutip/tests/core/test_coefficient.py:56]: Nitpick: Missing space between `+` and `t`.

File: qutip/tests/core/test_coefficient.py
Comment (hodgestar) [inline qutip/tests/core/test_coefficient.py:139]: We should just us fixed test cases here instead of random numbers, I think. If we use random numbers the test might just pass or fail randomly which is bad, and it doesn't really buy us anything over picking some specific numbers I don't think. If we want to test a range of different numbers, we should explicitly test the numbers we want to.

As an aside, I'd like to add hypothesis to qutip tests at some point, especially stuff in the core, but that's a bit of a rabbit hole, so I'll leave it until after I've started and version 5 has settled down a bit more.

**Review by hodgestar — COMMENTED** [review]
Thanks for breaking these out -- it was much easier to review this bit by itself.

I've left a few comments. I think Jake should weigh in too on these core things. One big picture concern is that quite a few changes have been made to the core here, but there are comparatively few tests that have been added or changed. We probably want to test these core things quite thoroughly if we can, since all of the rest of QuTiP relies on them.

File: qutip/core/cy/cqobjevo.pyx
Comment (Ericgig) [reply qutip/core/cy/cqobjevo.pyx:139]: It to not allocate memory when not needed.
`out = add(out, matmul(mat, vec))` (iadd is only available to Dense) would need to allocate 2 vectors for the operation, but none. In some case, I have seen this allocation being up to 30% of the time of the operation. 

We can work in-place for `Dense`'s `matmul`, but not `CSR`. So this method allow to skip any memory allocation for `Dense` and have a matching behaviour for other data type.

File: qutip/core/coefficient.py
Comment (Ericgig) [reply qutip/core/coefficient.py:355]: Good point, if the hash is not sufficient and we don't support infinite hash collision like `dict`, we should just quickly fail instead of trying to postpone. Expecting about 10000 coefficients used per users, 30 hex digit should be more than enough.

File: qutip/core/coefficient.py
Comment (Ericgig) [reply qutip/core/coefficient.py:465]: Let's just fix it now.

File: qutip/core/cy/cqobjevo.pyx
Comment (jakelishman) [reply qutip/core/cy/cqobjevo.pyx:139]: I'm also going to have a think about re-using allocation space within the `CSR` type for `add` and `matmul`.  I've got some general ideas with a two-part swapping buffer that's half fleshed out in my head at the moment.

File: qutip/core/cy/cqobjevo.pyx
Comment (jakelishman) [inline qutip/core/cy/cqobjevo.pyx:220]: All these calls happen with Python semantics - is that intended?

File: qutip/installsettings.py
Comment (jakelishman) [inline qutip/installsettings.py:61]: This is safe on Windows, right?

File: qutip/tests/core/test_coefficient.py
Comment (jakelishman) [reply qutip/tests/core/test_coefficient.py:139]: Nathan asked me about this around the time I first introduced pytest, but I didn't know how to do it at the time.  I was basically learning how to do _any_ unit testing (no formal training strikes again!).  Seems like a good idea to me in general - random testing has bitten us a few times, although I did make some efforts to try and make sure that pytest reported _which_ random values were used so it could be reproduced.

**Review by jakelishman — DISMISSED** [review]
LGTM in general - one comment about some questionable error handling, and a couple of questions about stuff I wasn't sure.

File: qutip/core/__init__.py
Comment (Ericgig) [reply qutip/core/__init__.py:6]: I am thinking of removing `QobjEvoFunc` in the close future. But otherwise it can be as visible as `QobjEvo`.

File: qutip/core/cy/cqobjevo.pyx
Comment (Ericgig) [reply qutip/core/cy/cqobjevo.pyx:220]: Could be made cpdef, but `self.base` is a python object so it probably have only a small impact.
And it won't survive the merger of QobjEvo and cQobjEvo.

File: qutip/core/coefficient.py
Comment (Ericgig) [reply qutip/core/coefficient.py:449]: Yeah.. and we are moving from root to root... The whole block does nothing.

**Comment by Ericgig** [comment]
@hodgestar Could you tell me if I have answered comment and approve this?

File: qutip/core/cy/cqobjevo.pyx
Comment (jakelishman) [reply qutip/core/cy/cqobjevo.pyx:220]: Yeah, the `self.base` was the bit I was asking about.  As long as it's expected that it's all Python semantics, that's fine.  I forgot which bits were meant to be fastcall and which weren't.

File: qutip/core/__init__.py
Comment (Ericgig) [reply qutip/core/__init__.py:6]: I wrote my plans for `QobjEvoFunc` in #1468.

Comment (jakelishman) [reply qutip/core/__init__.py:6]: Ok, I'm almost always in favour of simplifying public interfaces (Y)

File: qutip/installsettings.py
Comment (hodgestar) [reply qutip/installsettings.py:61]: https://docs.python.org/3/library/os.path.html#os.path.expanduser is documented as supporting Windows.

File: qutip/core/cy/cqobjevo.pyx
Comment (hodgestar) [reply qutip/core/cy/cqobjevo.pyx:139]: My point is that in this line:
```python
out = _data.add(out, _data.matmul(self.constant, matrix))
```
the `_data.matmul(self.constant, matrix)` part appears to already creates a temporary object, so we appear to not have saved any temporary objects over just returning `out` and allowing the caller to do `caller_out = _data.add(caller_out, out)` themselves (note that I don't propose having the caller call `_data.add` directly -- rather that they call it via some friendlier interface).

Do you agree? Or did I miss something about how this works?

File: qutip/tests/core/test_coefficient.py
Comment (hodgestar) [inline qutip/tests/core/test_coefficient.py:45]: We shouldn't call clean up on import (otherwise any code that imports this test file will clear the cache). Instead we could add it to a `setup_function` function (see https://docs.pytest.org/en/stable/xunit_setup.html). This would also mean that the cache would be cleared between each test which would improve the isolation of the tests from each other.

**Review by hodgestar — APPROVED** [review]
@Ericgig Thank you for addressing and replying to my previous comments. I left two more, but I think the overall shape of this PR is good, so I'm approving it. Happy for you to merge it once you're happy.

**Comment by jakelishman** [comment]
That test failure actually looks like a serious problem - there's a `NaN` that has crept into an output after a `copy` operation.  In this case I think it's important that we really track down the root cause of how that happened, so we can be sure it can't happen again.

**Comment by Ericgig** [comment]
I could not replicate the bug locally, but I have a good guess.
I think it's fixed in changing `t>` to `t>=` in inter.pyx.

**Comment by Ericgig** [comment]
I rechecked the bug and it's really the case bad handling the limit of the range.
At `t=tlist[-1]`, it effectively did `y[len(y)-1] + 0*y[len(y)]` with boundary unchecked.

**Comment by Ericgig** [comment]
@jakelishman If you are happy with the bug fix, can I merge this and the next PR and prepare the next batch?

**Comment by jakelishman** [comment]
Oh, sorry I missed this.  The problem you found sounds pretty likely to be the cause.  Let's keep it in mind, just in case.  This looks fine to me to merge.

Let me know if you're waiting on me on other PRs, or poke me (via GitHub "request review" or by email) if in the future I'm not replying after a few days...  I haven't quite figured out what I want my GitHub notification settings to be yet in order to prioritise the most important things.

**Review by jakelishman — APPROVED** [review]

File: qutip/tests/core/test_coefficient.py
Comment (hodgestar) [inline qutip/tests/core/test_coefficient.py:62]: Belated review: Probably these lines should be
```python
def _assert_eq_over_interval(...):
    """ assert coeff1 == coeff2
         inside refer to the range covered by tlistlog: [0.01, 1]
    """
```
or
```python
def _assert_eq_over_interval(...):
    """ assert coeff1 == coeff2 """
    # inside refer to the range covered by tlistlog: [0.01, 1]
```
As things currently stand the docstring is `"assert coeff1 == coeff2"` and then the `"inside ..."` is just a statement on a line by itself.

**Comment by hodgestar** [comment]
@Ericgig Thanks again for splitting this out. It was a big help. Apologies for being away for the last few commits.

**Pull request merged**
